### DISCUSSION TITLE
Is it possible to Limit or Sample Parquet file(s) via predicate pushdown?

### DISCUSSION THREAD

**Discussion by be-rock** [duckdb/duckdb#1688] [Q&A] [answered]
Hi,

I seem to be unable to take advantage of what the Parquet docs describe as "Partial Reading". The goal is to simply view and sample 5 records but the 42 second response time seems to suggest that *all* the data is being read/fetched and then subsequently applying predicates. Am I missing something?

```
import duckdb

PREFIX = "/path/to/parquet/files/*"

con = duckdb.connect(database=':memory:', read_only=False)
con.execute(f"CREATE VIEW test AS SELECT * FROM parquet_scan('{PREFIX}')")
con.execute("select count(*) from test limit 5")
print(con.fetchall())
```

```
$ time python duckdbtest.py
[(5,)]
python   42.32s user 78.82s system 98% cpu 2:02.90 total
```

So what appears to be happening is that:
1. All parquet data is read and fetched
2. And *then* filters are being applied

Instead of:
1. Pushdown filters/predicates to the datastore and only read/fetch what satisfies the predicate. The docs describe this as "Partial Reading" source: https://duckdb.org/docs/data/parquet but I am not seeing this in action. Is there a way to 'activate' this?

Thanks.

**Comment by Mytherin** [comment, marked as answer]
If you want to sample five rows I assume you might be intending to run `SELECT * FROM test LIMIT 5`, rather than `SELECT COUNT(*) ...`. The `COUNT(*)` forces the system to scan all the Parquet files. In SQL a limit is a result modifier and is applied *after* any aggregates are computed, and the aggregates themselves require access to all data of the table (unless filters are provided, which are not present in this example).

Depending on how many Parquet files you have, metadata reading might be taking up the majority of the 42 seconds as well. You may want to run `pragma enable_object_cache` to enable the Parquet metadata cache.

**Reply by be-rock** [reply]
Thanks for the response. I see the same performance with:

```
SELECT * FROM test LIMIT 5
```

I am just looking for a way to efficiently sample the data in the parquet files. There is a single 3gb file in the PREFIX.

Thanks again.

**Reply by Mytherin** [reply]
In principle the LIMIT should work for this just fine. I have just tested this myself and it does work (on the master branch, at least). What does your Parquet file look like? How many rows/columns are there? What are their types? How many row groups are there?

**Reply by be-rock** [reply]
Thanks. The file has 930 columns, I no longer have the original file but I have to believe it was less than 100k rows. When you say 'it does work', does that mean you are able to fetch 5 rows in a few seconds?

**Reply by hannes** [reply]
It's tricky to find out what happened without the file. My guess would be too many/small row groups. Can you recreate a similar file that demonstrates the issue?

**Reply by be-rock** [reply]
Understood, thank you for your time, i will probably not have another chance to look at this very soon but if I do I will share updates again. Thanks again for the great work on this awesome project.

**Reply by dforsber** [reply]
Reading all columns from Parquet file(s) seems to be slow in general and I think this relates to the fact that the data is stored in columnar format. 930 columns is quite a lot and it means every column needs to be fetched (at least one row group), decompressed (if it was compressed) and then the first 5 rows read from that. Every row group needs to be read from different part of the disk.

**Reply by dforsber** [reply]
You can see the speed difference if you select less columns instead of "*".
